documentation/gnmi/dash-gnmi-design.md

```diff
@@ -0,0 +1,144 @@
+# GNMI Interface for DASH
+
+## Overview
+
+SONiC should provide a gNMI server interface for DASH project.
+Below diagram shows the architecture of the DASH project, the host server and DPU card are connected through PCIE, and together they form a smart switch.
+
+![dash-arch](./images/gnmi-arch.svg)
+
+Both host server and DPU card are running SONiC image. GNMI server and redis for APPL_DB are on host server, orchagent is on DPU card, GNMI server will use ZMQ over TCP to communicate with orchagent and redis.
+
+### Requirements
+
+* Set and get RPCs must be supported. Customers will use get RPC to retrieve DASH configurations and use set PRC to apply new DASH configurations.
+* Need to configure huge DASH table entries to APPL_DB, with high speed.
+* Minimal redis table scaling requirements:
+
+|Item |Expected value |
+|--|--|
+|VNETs|1024|
+|ENI|64 per card|
+|Outbound Routes per ENI|100K|
+|Inbound Routes per ENI|10K|
+|NSGs per ENI|10|
+|ACL rules per NSG|1000|
+|ACL prefixed per ENI|10*100K|
+|Max prefixed per rule|8K|
+|ACL ports per ENI|10*10K SRC/DST ports|
+|CA-PA mapping|10M per card|
+
+### Design considerations
+
+* For set RPC, GNMI calls the ZMQ interface, and ZMQ updates redis asynchronously.
+* For get RPC, GNMI reads from redis db directly.
+* The DASH table of APPL_DB would be encoded as protobuf to save memory consumption, and then GNMI needs to support protobuf encoding for DASH table.
+
+```
+127.0.0.1:6379> hgetall "DASH_VNET_TABLE:vnet1"
+1\) "pb"
+2\) "\n\x010\x12$b6d54023-5d24-47de-ae94-8afe693dd1fc…"
+```
+
+And proto message for DASH_VNET_TABLE is:
+
+```
+message Vnet {
+    string vni = 1;
```

Comment:
I changed the type from string to uint32 for vni.
https://github.com/sonic-net/DASH/pull/350/files#diff-fb906e1fa18c78bfd38451e4a8de66fe3ff0194cfc466a21c3772286629867d3R710

Reply:
updated.